File: python/quicksort.py
import random

"""
Implementation of recursive quick sort.
"""


def quicksort(arr):
    if len(arr) <= 1:
        return arr

    first, pivot, last = partition(arr)
    result = quicksort(first) + [pivot] + quicksort(last)
    return result


def partition(arr):
    # randomly select pivot
    arr_size = len(arr)
    pivot_index = random.randrange(arr_size)
    pivot = arr[pivot_index]

    # move pivot to head of the array
    tmp = arr[0]
    arr[0] = arr[pivot_index]
    arr[pivot_index] = tmp

    i = 0
    j = arr_size - 1
    while i <= j:
        # 이렇게 하면 마지막 단계에서 무조건 i - 1번째 element가 pivot보다 같거나 작다
        # i가 먼저 움직이기 때문
        while i <= j and arr[i] <= pivot:
            i += 1

        while i <= j and arr[j] > pivot:
            j -= 1

        if i <= j:
            # swap
            tmp = arr[i]
            arr[i] = arr[j]
            arr[j] = tmp

    # pivot을 가운데로 갖다놓는다
    tmp = arr[0]
    arr[0] = arr[i - 1]
    arr[i - 1] = tmp
    return arr[:i - 1], arr[i - 1], arr[i:]


if __name__ == '__main__':
    array = [1, 2, 6, 4, 267, 8, 6, 4, 2, 5, 11, 99, 6, 3]
    print(quicksort(array))

File: java/src/QuickSort.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;


class QuickSort {
  public static void swap(List<Integer> arr, int i, int j) {
    int tmp = arr.get(i);
    arr.set(i, arr.get(j));
    arr.set(j, tmp);
  }

  public static int partition(List<Integer> arr, int left, int right) {
    int mid = (left + right) / 2;
    int pivot = arr.get(mid);
    swap(arr, mid, left);  // save the pivot value somewhere

    int i = left;
    int j = right;
    while (i <= j) {
      if (arr.get(i) <= pivot && i <= j) {
        i++;
      }

      if (arr.get(j) > pivot && i <= j) {
        j--;
      }

      if (i <= j) {
        swap(arr, i, j);
      }
    }

    swap(arr, i - 1, left);
    return i - 1; // this is the position of the pivot
  }

  public static void sort(List<Integer> arr, int left, int right) {
    if (left > right) {
      return;
    }

    int part = partition(arr, left, right);
    sort(arr, left, part - 1);
    sort(arr, part + 1, right);
  }

  public static void main(String[] args) throws java.io.IOException {
    BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));  // faster read
    int n = Integer.parseInt(bf.readLine());

    List<Integer> arr = new ArrayList<Integer>();
    for (int i = 0; i < n; i++) {
      arr.add(Integer.parseInt(bf.readLine()));
    }

    sort(arr, 0, arr.size() - 1);
    for (int j = 0; j < n; j++) {
      System.out.println(arr.get(j));
    }
  }
}
